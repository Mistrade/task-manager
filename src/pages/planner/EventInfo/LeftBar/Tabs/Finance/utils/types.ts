import { MRT_FilterFns } from 'material-react-table';

import {
  FINANCE_OPERATION_TYPES,
  IFinanceOperation,
} from '@api/finance-api/types';

type FilterNumberFields = keyof Pick<
  IFinanceOperation,
  'result' | 'count' | 'value'
>;
type FilterStringFields = keyof Pick<
  IFinanceOperation,
  'operationType' | 'name'
>;
type FilterDateFields = keyof Pick<IFinanceOperation, 'date'>;
export type FilterFields =
  | FilterNumberFields
  | FilterStringFields
  | FilterDateFields;

type AgreementFns = Pick<
  typeof MRT_FilterFns,
  | 'lessThanOrEqualTo'
  | 'lessThan'
  | 'equals'
  | 'includesString'
  | 'greaterThanOrEqualTo'
  | 'greaterThan'
  | 'equalsString'
  | 'startsWith'
  | 'endsWith'
>;
export type AgreementNumberFns = Pick<
  AgreementFns,
  | 'lessThanOrEqualTo'
  | 'lessThan'
  | 'equals'
  | 'greaterThanOrEqualTo'
  | 'greaterThan'
>;

export type AgreementDateFns = AgreementNumberFns;

export type AgreementStringFns = Pick<
  AgreementFns,
  'equalsString' | 'includesString' | 'startsWith' | 'endsWith'
>;

type TColumnsNumbersFns = {
  [key in FilterNumberFields]: keyof AgreementNumberFns;
};

type TColumnsStringFns = {
  [key in FilterStringFields]: keyof AgreementStringFns;
};

type TColumnDateFns = {
  date: keyof AgreementDateFns;
};

export type TColumnsFilterFnsMap = TColumnsNumbersFns &
  TColumnsStringFns &
  TColumnDateFns;

type TColumnFilterNumberValue = {
  id: FilterNumberFields;
  value: number;
};

type TColumnFilterStringValues = {
  id: keyof Pick<IFinanceOperation, 'name'>;
  value: string;
};

type TColumnsFilterOperationType = {
  id: keyof Pick<IFinanceOperation, 'operationType'>;
  value: FINANCE_OPERATION_TYPES;
};

type TColumnsFilterDateValues = {
  id: keyof Pick<IFinanceOperation, 'date'>;
  value: Date;
};

type TFilterCallbackNumberValues = {
  [key in TColumnFilterNumberValue['id']]: TColumnFilterNumberValue['value'];
};

type TFilterCallbackStringValues = {
  [key in TColumnFilterStringValues['id']]: TColumnFilterStringValues['value'];
};

type TFilterCallbackEnumValues = {
  [key in TColumnsFilterOperationType['id']]: TColumnsFilterOperationType['value'];
};

type TFilterCallbackDateValues = {
  [key in TColumnsFilterDateValues['id']]: TColumnsFilterDateValues['value'];
};

type TColumnFilterValues =
  | TColumnFilterNumberValue
  | TColumnFilterStringValues
  | TColumnsFilterOperationType
  | TColumnsFilterDateValues;

export type TColumnFilterValuesMap = {
  count: TColumnFilterNumberValue;
  value: TColumnFilterNumberValue;
  result: TColumnFilterNumberValue;
  name: TColumnFilterStringValues;
  operationType: TColumnsFilterOperationType;
  date: TColumnsFilterDateValues;
};

type TFilterCallbackValues = TFilterCallbackNumberValues &
  TFilterCallbackStringValues &
  TFilterCallbackEnumValues &
  TFilterCallbackDateValues;

export interface IColumnFilterState {
  fns: TColumnsFilterFnsMap;
  values: Array<TColumnFilterValues>;
}

type TCallbackValues<T extends FilterFields> = {
  [key in T]: {
    key: T;
    pattern: TColumnsFilterFnsMap[T];
    value: TFilterCallbackValues[T] | null;
  };
};

export type TCallbackData<Key extends FilterFields> = TCallbackValues<Key>[Key];
