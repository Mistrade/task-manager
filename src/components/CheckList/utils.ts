import {ICheckListItem} from "@planner/types";
import dayjs from "dayjs";
import {DATE_RENDER_FORMAT} from "../../common/constants/defaultConstants";

export function generateCheckListCopyContent (title: string, arr: Array<ICheckListItem>): string {
	const resultTitle = `Чек-лист: ${title}`;
	const items = arr.map((item, index) => `${index + 1}. ${item.title}`).join('\n');
	const copyright = `Сгенерировано сервисом WhitePlanner ${dayjs().format(DATE_RENDER_FORMAT)}`
	return `${resultTitle}\n\n${items}\n\n${copyright}`
}