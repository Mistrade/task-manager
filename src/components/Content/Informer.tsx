import styled from "styled-components";
import {ErrorTypes} from "../../store/api/taskApi/taskApi";
import React, {FC, ReactNode} from "react";

interface DefaultInformerProps {
	type: ErrorTypes
}

export const StyledInformer = styled('div')<DefaultInformerProps>`
  display: flex;
  padding: 8px;
	flex-direction: column;
	gap: 8px;
`

export const StyledInformerHeader = styled('h4')`
  width: 100%;
  font-size: 17px;
  font-weight: bold;
  font-style: normal;
  font-family: Helvetica, sans-serif;
`

export const StyledInformerText = styled('p')`
  width: 100%;
  font-size: 14px;
  font-weight: 400;
  font-style: normal;
  font-family: Helvetica, sans-serif;
`

interface InformerProps extends DefaultInformerProps {
	header?: ReactNode,
	text: ReactNode,
	"data-testid"?: any  //todo
}

export const Informer: FC<InformerProps> = ({type, header, text}) => {
	return (
		<StyledInformer type={type}>
			{header && (
				<StyledInformerHeader>
					{header}
				</StyledInformerHeader>
			)}
			<StyledInformerText>
				{text}
			</StyledInformerText>
		</StyledInformer>
	)
}