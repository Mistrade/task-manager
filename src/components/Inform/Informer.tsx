import {
  borderRadiusSize,
  currentColor,
  hoverColor,
} from '../../common/constants';
import { FlexBlock } from '../LayoutComponents/FlexBlock';
import { FC, ReactNode } from 'react';

export const Informer: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <FlexBlock
      mb={12}
      justify={'flex-start'}
      width={'100%'}
      maxWidth={'100%'}
      borderRadius={borderRadiusSize.xs}
      p={16}
      border={`1px solid ${currentColor}`}
      bgColor={hoverColor}
    >
      {children}
    </FlexBlock>
  );
};
