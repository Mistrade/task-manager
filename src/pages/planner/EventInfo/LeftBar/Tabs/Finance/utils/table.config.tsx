import { DateHelper } from '../../../../../../../common/calendarSupport/dateHelper';
import { FINANCE_OPERATION_TITLES } from '../components';
import {
  FINANCE_OPERATION_TYPES,
  IFinanceOperation,
  TFinanceOperationWithDate,
} from '@api/finance-api/types';
import { EventInfoModel } from '@api/planning-api/types/event-info.types';
import { Row } from '@tanstack/table-core/src/types';
import React from 'react';


export function sortingNumberColumns(
  rowA: Row<IFinanceOperation>,
  rowB: Row<IFinanceOperation>,
  key: 'count' | 'value' | 'result'
): number {
  const a = rowA.original[key];
  const b = rowB.original[key];

  return a - b;
}

export function getFinanceOperationValue(num: number): string {
  let value = `${num}`.split('');
  const arr = [];

  while (value.length) {
    const arrSym = value.splice(-3, 3);
    arr.unshift(arrSym.join(''));
  }

  return arr.join(' ').trim() + ' ₽';
}

console.log('test split number: ', getFinanceOperationValue(10000));

export function buildEventDescriptionFromOperation(
  original: TFinanceOperationWithDate,
  eventInfo: EventInfoModel
) {
  return `Событие было создано от финансовой операции \"${
    original.name
  }\", закрепленной за событием "${eventInfo.title}".
Информация по финансовой операции:
Название - ${original.name}
Тип - ${FINANCE_OPERATION_TITLES[original.operationType]},
Дата - ${
    original.date
      ? DateHelper.getHumanizeDateValue(original.date, {
          withTime: true,
        })
      : 'не записана'
  },
Сумма - ${original.value},
Количество - ${original.count},
Результат - ${original.result}.`;
}

interface IGetBottomToolbarAnalyticReturned {
  profit: number;
  consumption: number;
  income: number;
  incomeCount: number;
  consumptionCount: number;
  allCount: number;
  maxIncome: TFinanceOperationWithDate | null;
  maxConsumption: TFinanceOperationWithDate | null;
  isPositiveResult: boolean;
}

export function getBottomToolbarAnalytic(
  rows: Array<{ original: TFinanceOperationWithDate }>
): IGetBottomToolbarAnalyticReturned {
  let profit = 0;
  let consumption = 0;
  let income = 0;
  let incomeCount = 0;
  let consumptionCount = 0;
  let allCount = 0;
  let maxIncome: TFinanceOperationWithDate | null = null;
  let maxConsumption: TFinanceOperationWithDate | null = null;

  rows.forEach((item, index, array) => {
    if (item.original.operationType === FINANCE_OPERATION_TYPES.INCOME) {
      profit += item.original.result;
      income += item.original.result;
      incomeCount++;

      if (maxIncome) {
        if (item.original.result > maxIncome.result) {
          maxIncome = item.original;
        }
      } else {
        maxIncome = item.original;
      }
    } else {
      profit -= item.original.result;
      consumption += item.original.result;
      consumptionCount++;

      if (maxConsumption) {
        if (item.original.result > maxConsumption.result) {
          maxConsumption = item.original;
        }
      } else {
        maxConsumption = item.original;
      }
    }

    allCount++;
  });

  const isPositiveResult = profit >= 0;

  return {
    profit,
    consumption,
    income,
    incomeCount,
    consumptionCount,
    allCount,
    maxIncome,
    maxConsumption,
    isPositiveResult,
  };
}