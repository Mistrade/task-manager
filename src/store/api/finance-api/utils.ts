import { ObjectId } from '../rtk-api.types';
import {
  FINANCE_OPERATION_TYPES,
  IFinanceModel,
  TOperationSwitchItem,
} from './types';

export function updateGetFinanceModelsData(initial: Array<IFinanceModel>) {
  const switcherData: Array<TOperationSwitchItem> = [];
  const models: { [key in ObjectId]: IFinanceModel } = {};

  initial?.forEach((item) => {
    switcherData.push({
      title: item.title,
      type: item._id,
      badgeValue: item.analytic.profit,
      badgeType:
        item.analytic.profit >= 0
          ? FINANCE_OPERATION_TYPES.INCOME
          : FINANCE_OPERATION_TYPES.CONSUMPTION,
    });
    models[item._id] = item;
  });

  return {
    switcherData,
    models,
  };
}
