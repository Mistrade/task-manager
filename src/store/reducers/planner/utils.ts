import { ObjectId } from '@api/rtk-api.types';
import { EventFilterTaskStatuses } from '@pages/planner/RenderModes/FindEventFilter/find-event-filters.types';
import { EVENT_INFORMER_TAB_NAMES } from '@pages/planner/TaskInformer/LeftBar/TaskInformerLeftBar';
import { IPlannerDate } from '@planner-reducer/types';
import { ServicesNames } from '@redux/reducers/global';
import { PLANNER_LAYOUTS, URLTaskStatuses } from '@src/common/constants';
import dayjs, { Dayjs } from 'dayjs';

export const isEqualPlannerDate = (
  prev: IPlannerDate,
  next: IPlannerDate
): boolean => {
  const dayIsEqual = prev.day === next.day;
  const monthIsEqual = prev.month === next.month;
  const yearIsEqual = prev.year === next.year;

  return dayIsEqual && monthIsEqual && yearIsEqual;
};

export const dateToPlannerDate = (
  date: Date | Dayjs | string
): IPlannerDate => {
  const value = date instanceof Dayjs ? date : dayjs(date);

  return {
    day: value.date(),
    month: value.month(),
    year: value.year(),
  };
};

export const isCorrectLayout = (value: string): boolean => {
  const arrOfLayouts: Array<string> = Object.values(PLANNER_LAYOUTS);
  return arrOfLayouts.includes(value);
};

export const isCorrectStatus = (value: string): boolean => {
  return Object.keys(URLTaskStatuses).includes(value);
};

export const isCorrectEventId = (value: string): boolean => {
  return value.length === 24;
};

export const isCorrectEventInformerTabName = (value: string): boolean => {
  const arr: Array<string> = Object.values(EVENT_INFORMER_TAB_NAMES);
  return arr.includes(value);
};

export const getPlannerMetaData = (
  path: string
): {
  layout: PLANNER_LAYOUTS | null;
  status: EventFilterTaskStatuses | null;
  eventInfoId: ObjectId | null;
  eventInformerTabName: EVENT_INFORMER_TAB_NAMES | null;
} => {
  const [serviceName, layout, status, eventPath, infoPath, eventId, tabName] =
    path.split('/').filter((part) => part.length);

  if (serviceName === ServicesNames.PLANNER) {
    return {
      layout:
        layout && isCorrectLayout(layout) ? (layout as PLANNER_LAYOUTS) : null,
      status:
        status && isCorrectStatus(status)
          ? (status as EventFilterTaskStatuses)
          : null,
      eventInfoId:
        eventId && isCorrectEventId(eventId) ? (eventId as ObjectId) : null,
      eventInformerTabName:
        tabName && isCorrectEventInformerTabName(tabName)
          ? (tabName as EVENT_INFORMER_TAB_NAMES)
          : null,
    };
  }
  return {
    layout: null,
    status: null,
    eventInfoId: null,
    eventInformerTabName: null,
  };
};
