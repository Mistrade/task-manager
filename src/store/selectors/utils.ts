import {PlannerMode} from "../../pages/Planner/planner.types";
import {CalendarModeForState} from "../reducers/planner-reducer";

export const currentFromStoreToDefault = (current: CalendarModeForState): PlannerMode => {
	const {layout} = current
	
	switch (layout) {
		case "day":
			return {
				layout: "day",
				date: new Date(current.date)
			}
		case "week":
			return {
				layout: 'week',
				aroundDate: new Date(current.aroundDate)
			}
		case "list":
			return {
				layout: 'list',
				fromDate: new Date(current.fromDate),
				toDate: new Date(current.toDate)
			}
		default:
			return current
	}
}