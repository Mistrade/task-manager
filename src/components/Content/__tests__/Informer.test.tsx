// import { render } from "react-dom";
import { TestRender } from "../../../utils/test-utils"
import { Informer, StyledInformer, StyledInformerHeader, StyledInformerText } from "../Informer";
import { render, screen } from "@testing-library/react";

describe('All Informer TestGroup', () => {
    //todo
    // it('test -StyledInformer', () => {
    //     const testId = 'StyledInformerTestId'
    //     const { container, getByTestId } = TestRender(
    //         <Informer text={undefined} type={"default"} data-testid={testId}>
    //         </Informer >
    //     )

    //     const test = getByTestId(testId)
    //     expect(test).toBeInTheDocument()
    //     expect(test.tagName.toLowerCase()).toBe('div')
    // })

    it('test-StyledInformer', () => {
        const testId = 'StyledInformer'
        const testChildren = 'Test-text'
        const header = "header-text"
        const text = "testing-message"
        const { getByTestId, getByText } = TestRender(
            <StyledInformer
                data-testid={testId}
                type="info"
            >
                {header && (
                    <StyledInformerHeader>
                        {header}
                    </StyledInformerHeader>
                )}
                <StyledInformerText>
                    {text}
                </StyledInformerText>
            </StyledInformer>
        )
        const test = getByTestId(testId)
        expect(test).toBeInTheDocument()
        expect(getByText(header)).toBeInTheDocument()
        expect(getByText(text)).toBeInTheDocument()
    })


    it('test-StyledInformerHeader', () => {
        const testId = 'StyledInformerHeader'
        const testChildren = 'TEST-text'
        const { getByTestId, getByText } = TestRender(
            <StyledInformerHeader data-testid={testId}>
                {testChildren}
            </StyledInformerHeader>
        )

        const test = getByTestId(testId)
        expect(test).toBeInTheDocument()
        expect(getByText(testChildren)).toBeInTheDocument()
        expect(test.tagName.toLowerCase()).toBe('h4')
    })

    it('test-StyledInformerText', () => {
        const testId = 'StyledInformerText'
        const testChildren = 'test-text'
        const { container, getByTestId, getByText } = TestRender(
            <StyledInformerText data-testid={testId}>
                {testChildren}
            </StyledInformerText>
        )
        const test = getByTestId(testId)
        expect(container).toMatchSnapshot()
        expect(test).toBeInTheDocument()
        expect(getByText(testChildren)).toBeInTheDocument()
        expect(test.tagName.toLowerCase()).toBe('p')

        // const { container } = render(<StyledInformerText>{testChildren}</StyledInformer>>)
    })

    //todo

    it('render StyledInformerHeader', () => {
        const testChildren = 'test-text'
        render(<StyledInformerHeader>{testChildren}</StyledInformerHeader>)
        //todo
        // expect(screen.getByRole('h4')).toBeInTheDocument()
        expect(screen.getByText(testChildren)).toBeInTheDocument()
    })

    it('render StyledInformerText', () => {
        const testChildren = 'test-text'
        render(<StyledInformerText>{testChildren}</StyledInformerText>)
        const linkElement = screen.getByText(testChildren);
        expect(linkElement).toBeInTheDocument();
    })
})
