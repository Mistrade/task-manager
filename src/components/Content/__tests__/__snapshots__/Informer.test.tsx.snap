// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Informer TestGroup test-StyledInformerText 1`] = `
<div>
  <p
    class="sc-hKMtZM jOypbF"
    data-testid="StyledInformerText"
  >
    test-text
  </p>
</div>
`;
